gapis/api/vulkan/memory_layout.go

```diff
@@ -0,0 +1,114 @@
+// service.MemoryLayout object.  The layout includes data on memory types,
+// allocations, and which resources are bound to which locations in memory.
+func (a API) MemoryLayout(st *api.GlobalState) (*service.MemoryLayout, error) {
+	s := GetState(st)
+	allocations := []*service.MemoryAllocation{}
+	// Serialize data on all allocations into protobufs
+	for handle, info := range *(s.DeviceMemories().Map) {
+		device := info.Device()
+		typ := info.MemoryTypeIndex()
+		flags, err := s.getMemoryTypeFlags(device, typ)
+		if err != nil {
+			return nil, err
+		}
+		bindings, err := s.getAllocationBindings(info.Get())
+		if err != nil {
+			return nil, err
+		}
+
+		mapping := service.MemoryMapping{
+			Size:        uint64(info.MappedSize()),
+			Offset:      uint64(info.MappedOffset()),
+			HostAddress: uint64(info.MappedLocation()),
+		}
+
+		alloc := service.MemoryAllocation{
+			Device:     uint64(info.Device()),
+			MemoryType: uint32(typ),
+			Flags:      uint32(flags),
+			Id:         uint64(handle),
+			Name:       strconv.FormatUint(uint64(handle), 10),
+			Size:       uint64(info.AllocationSize()),
+			Mapping:    &mapping,
+			Bindings:   bindings,
+		}
+
+		allocations = append(allocations, &alloc)
+	}
+	return &service.MemoryLayout{
+		Api:                  path.NewAPI(id.ID(ID)),
+		AllocationFlagsIndex: int32(VkMemoryPropertyFlagBitsConstants()),
+		Allocations:          allocations,
+	}, nil
+}
+
+func (s *State) getMemoryTypeFlags(device VkDevice, typeIndex uint32) (VkMemoryPropertyFlags, error) {
+	deviceObject := s.Devices().Get(device)
+	if deviceObject.IsNil() {
+		return VkMemoryPropertyFlags(0), fmt.Errorf("Failed to find device %v", device)
+	}
+	physicalDevice := deviceObject.PhysicalDevice()
+	physicalDeviceObject := s.PhysicalDevices().Get(physicalDevice)
+	if physicalDeviceObject.IsNil() {
+		return VkMemoryPropertyFlags(0), fmt.Errorf("Failed to find physical device %v", physicalDevice)
+	}
+	props := physicalDeviceObject.MemoryProperties()
+	if props.MemoryTypeCount() <= typeIndex {
+		return VkMemoryPropertyFlags(0), fmt.Errorf("Memory type %v is larger than physical device %v's number of memory types (%v)",
+			typeIndex, physicalDevice, props.MemoryTypeCount())
+	}
+	return props.MemoryTypes().Get(int(typeIndex)).PropertyFlags(), nil
+}
+
+func (s *State) getAllocationBindings(allocation DeviceMemoryObject) ([]*service.MemoryBinding, error) {
+	bindings := []*service.MemoryBinding{}
+	for handle, offset := range allocation.BoundObjects().All() {
+		var size VkDeviceSize
+		var typ service.MemoryBinding_BindingType
+		if buffer, ok := s.Buffers().Lookup(VkBuffer(handle)); ok {
+			typ = service.MemoryBinding_BUFFER
+			size = buffer.Info().Size()
+		} else if image, ok := s.Images().Lookup(VkImage(handle)); ok {
+			typ = service.MemoryBinding_IMAGE
+			size = image.MemoryRequirements().Size()
+		} else {
+			return nil, fmt.Errorf("Bound object %v is not a buffer or an image", handle)
```

Comment:
I remember for sparse image and buffers, we didn't register the binding info on the memory object, only the buffer/image object. So we probably want to have a TODO or Note here to say some cases are not handled yet.